### ISSUE TITLE
Understanding the conversion operations and generating the rotation quaternion from a 3d cuboid.

### ISSUE THREAD

**Issue by Nicholas-Mitchell** [moble/quaternion#149]
First of all, thanks for this amazing package!

The TL;DR is that I would like to understand why converting a quaternion to a rotation matrix, then back to a quaternion does not reproduce the input. Here is a working example:

```python
import numpy as np
import quaternion

# Starting point: a float array
float_quat = np.array([0.70033807, 0.0014434, -0.00209336, 0.71380675], dtype=np.float32)

# Make the quaternion object
q = quaternion.from_float_array(float_quat)
# print(q) ->   quaternion(0.700338065624237, 0.00144340004771948, -0.00209336006082594, 0.713806748390198)

# Generate the rotation matrix
rot = quaternion.as_rotation_matrix(q)
# print(rot)
# array([[-1.90489677e-02, -9.99818172e-01, -8.71502129e-04],
#        [ 9.99806086e-01, -1.90443702e-02, -5.01024534e-03],
#        [ 4.99273713e-03, -9.66773134e-04,  9.99987069e-01]])

# Go back to the quaternion representation
q_from_rot = quaternion.from_rotation_matrix(rot)
# print(q_from_rot)
# quaternion(-0.700338084614, -0.00144340008685747, 0.00209336011758765, -0.713806767745166)
```
The only difference is that all the signs have been reversed (neglecting difference due to precision).

----

I am working on a problem whereby I receive 3d objects of the form: centroids _(x, y, z)_, dimensions _(length, width, height)_ and quaternion-rotation _(w, x, y, z)_.
For my purposes, I must represent these objects as 3d cuboids and (for speed) rotation matrices. After the processing, I need to once again restore a representation of centroids and quaternions, based on the 8 points of the 3d cuboid.

I thought if I can manually create the rotation matrix from the final cuboid, I could use `quaternion.from_rotation_matrix()` to get the final quaternion representation (note that I don't actually expect the same quaternion that I began with, as the object may have been rotated). I started out by testing the conversions as in the code above, from a rotation matrix to quaternion.

Is there another way to achieve this that I am overlooking?

Thanks in advance for any help!

**Comment by willwray** [comment]
> ... why converting a quaternion to a rotation matrix, then back to a quaternion does not reproduce the input.
> ... The only difference is that all the signs have been reversed

It is an inherent property; the unit quaternions 'double-cover' the space of 3D rotations.
This is covered in most references, *Rotations, Quaternions and Double Groups*
 by Simon L Altmann has maybe the most extensive discussion (not a recommendation).

Unit quaternions q and -q correspond to the same rotation matrix R.
The quaternion is effectively a square root of the rotation represented by R.
Or, vice-versa, R is effectively the square of q.

The formula for R in quaternion components w,x,y,z involves only order 2 terms:
```
R = np.array([
        [w*w + x*x - y*y - z*z,   2*(x*y - w*z),       2*(z*x + w*y)    ],
        [  2*(x*y + w*z),     w*w - x*x + y*y - z*z,   2*(y*z - w*x)    ],
        [  2*(z*x - w*y),       2*(y*z + w*x),     w*w - x*x - y*y + z*z]
    ])
```
so it is clear that changing the sign of q has no effect on R.
In the absence of other information the choice of sign is arbitrary.

This double-cover property can be useful or essential in some applications.
It can be a nuisance in others.

In your case, if you still have the original input quaternion q0, and know that
further processing of R will not rotate it by more than a full turn, 2*pi,
then you can choose the sign of the resulting quaternion q1 to maximize the
'scalar product' of the quaternions q0 * q1.conjugate()

**Comment by moble** [comment]
Will's comment is correct.  The quaternions `q` and `-q` are different quaternions, but *they represent the same rotation*.  You can see this in the way a quaternion is used to rotate a vector `v`: the rotated vector is `q * v * q.conjugate()`, which is precisely the same as `(-q) * v * (-q).conjugate()`.  When you convert to a rotation matrix, you "forget" that sign information, and there's no way to regain it from a rotation matrix alone.

But this doesn't mean you can always just ignore the difference.  In particular, if you're looking at a series of quaternions (say, as functions of time), then you need them to be reasonably continuous — and in particular, not randomly flipping signs.  If you were to try to `slerp` or `squad` interpolate between "nearby" rotations represented by quaternions with opposite signs, you would get some pretty crazy orientations in between.

Generally, there are two possible solutions if the signs actually matter.  The first would be to never transform away from quaternions — though you say you do this for speed.  I don't know anything about this "processing" step you're doing, so I can't suggest alternatives, but it might be worth thinking about and/or testing to ensure that there really is a significant speed difference.

The second solution is — as Will suggests — to apply your own sign flips to make things more consistent.  However, I would suggest a different condition than Will suggests.  Using `(q0 * q1.conjugate()).w` requires a full quaternion multiplication, only to throw away 3 components of the result.  (And doing it manually is even slower, because python.)  It's a bit more efficient to just use the absolute value of the difference.  So you could use a function something like this:

```python
def sign_corrected(q0, q1):
    """Adjust the sign of q1 to be as close as possible to q0"""
    if abs(q0 - q1) > 1.4142135623730951:
        return -q1
    return q1
```

---

Also, a brief note just to uphold the reputation of this package :) : If you normalize `q` by defining it as `q = quaternion.from_float_array(float_quat).normalized()` in the original code, this means that it actually represents a rotation, and the difference between `q` and `-q_from_rot` is just 3e-16, which is essentially the limit of accuracy in double-precision numerics — and therefore, about as good as we could ever hope to do.

**Comment by Nicholas-Mitchell** [comment]
Thank you @willwray and @moble for both of your explanations; it really really helps.

One small note on the perhaps _random_ change of sign: if I force the function to use the simpler implementation by setting `nonorthogonal=False` (and thereby not using `scipy.linalg`), I do get back the same signs as the original quaternion in all my tests. Perhaps it was a a fluke. I should be able to hold the value of the original quaternion, so could definitely use your small helper function to maintain the final sign.

There are several parts operations I perform on a set of 3d cuboids, many of which require an 8-point bounding box representation for each cuboid. I generate that representation via the rotation matrix `R = quaternion.as_rotation_matrix(q)`, applying it to the axis-aligned cuboid that I compute from its centroid and dimensions. Not being _fluent_ with quaternions, It was not clear to me how to 1) rotate the 8 points on the cuboid's own axis, retrieving the correct orientation of the 8-point cuboid, and 2) rotate the correctly oriented cuboid around the axes of the scene; i.e. the cuboids are each translated from the origin.

Generating the cuboids is something I can do relatively quickly, e.g. in ~130 μs, using vectorised numpy and _numba_'s `@jit` decorator, so if I don't take a performance hit by sticking to quaternions, it could be feasible to generate the cuboids multiple times, as and when required.

If this sounds like something that is easily achievable, I'd appreciate any pointers you could provide :)   I can provide a small sample dataset, if my description is a little fuzzy.

**Comment by moble** [comment]
> if I force the function to use the simpler implementation ..., I do get back the same signs as the original quaternion in all my tests. Perhaps it was a a fluke.

It's definitely a fluke.  The resulting signs aren't truly random, but they're determined by such low-level details that it's a waste of time trying to read any significance into them.  The bottom line is that — at a fundamental mathematical level — there's no "correct" answer for the question of which sign the quaternion output by `from_rotation_matrix` should have.

> It was not clear to me how to 1) rotate the 8 points on the cuboid's own axis, retrieving the correct orientation of the 8-point cuboid, and 2) rotate the correctly oriented cuboid around the axes of the scene; i.e. the cuboids are each translated from the origin.

I'm not sure what's unclear to you, but I'll point out that [the composition of rotations is given by multiplying the corresponding quaternions.](https://en.wikipedia.org/wiki/Quaternions_and_spatial_rotation#The_composition_of_spatial_rotations)  And you can also rotate a displacement vector by a quaternion.

> I'd appreciate any pointers you could provide

At least for the part you've described here, you could do better by explicitly calculating the coordinates of the cube's vertices directly.  For example, if you start with a reference cube centered on the origin with vertices given by the 8x3 array

```python
vertices = np.array([[x, y, z] for x in [-1, 1] for y in [-1, 1] for z in [-1, 1]])
```

and you have to generate a bunch of rotated copies of that cube from some Nx4 array `rotors` (where the function `quaternion.as_float_array` might be helpful), then — assuming the rotors are normalized — this function will return an Nx8x3 array of the coordinates of the vertices of those cubes:

```python
@numba.jit
def rotate_many_cubes(rotors):
    new_vertices = np.empty((rotors.shape[0], 8, 3))
    for i in range(rotors.shape[0]):
        r0, r1, r2, r3 = rotors[i, 0], rotors[i, 1], rotors[i, 2], rotors[i, 3]
        new_vertices[i, 0, 0] = r0*(-r0 - 2*r2 + 2*r3) + r1*(-r1 - 2*r2 - 2*r3) + r2**2 + r3**2
        new_vertices[i, 0, 1] = r0*(-r0 + 2*r1 - 2*r3) + r1*(r1 - 2*r2) + r2*(-r2 - 2*r3) + r3**2
        new_vertices[i, 0, 2] = r0*(-r0 - 2*r1 + 2*r2) + r1*(r1 - 2*r3) + r2*(r2 - 2*r3) - r3**2
        new_vertices[i, 1, 0] = r0*(-r0 + 2*r2 + 2*r3) + r1*(-r1 - 2*r2 + 2*r3) + r2**2 + r3**2
        new_vertices[i, 1, 1] = r0*(-r0 - 2*r1 - 2*r3) + r1*(r1 - 2*r2) + r2*(-r2 + 2*r3) + r3**2
        new_vertices[i, 1, 2] = r0*(r0 - 2*r1 + 2*r2) + r1*(-r1 - 2*r3) + r2*(-r2 - 2*r3) + r3**2
        new_vertices[i, 2, 0] = r0*(-r0 - 2*r2 - 2*r3) + r1*(-r1 + 2*r2 - 2*r3) + r2**2 + r3**2
        new_vertices[i, 2, 1] = r0*(r0 + 2*r1 - 2*r3) + r1*(-r1 - 2*r2) + r2*(r2 - 2*r3) - r3**2
        new_vertices[i, 2, 2] = r0*(-r0 + 2*r1 + 2*r2) + r1*(r1 - 2*r3) + r2*(r2 + 2*r3) - r3**2
        new_vertices[i, 3, 0] = r0*(-r0 + 2*r2 - 2*r3) + r1*(-r1 + 2*r2 + 2*r3) + r2**2 + r3**2
        new_vertices[i, 3, 1] = r0*(r0 - 2*r1 - 2*r3) + r1*(-r1 - 2*r2) + r2*(r2 + 2*r3) - r3**2
        new_vertices[i, 3, 2] = r0*(r0 + 2*r1 + 2*r2) + r1*(-r1 - 2*r3) + r2*(-r2 + 2*r3) + r3**2
        new_vertices[i, 4, 0] = r0*(r0 - 2*r2 + 2*r3) + r1*(r1 - 2*r2 - 2*r3) - r2**2 - r3**2
        new_vertices[i, 4, 1] = r0*(-r0 + 2*r1 + 2*r3) + r1*(r1 + 2*r2) + r2*(-r2 - 2*r3) + r3**2
        new_vertices[i, 4, 2] = r0*(-r0 - 2*r1 - 2*r2) + r1*(r1 + 2*r3) + r2*(r2 - 2*r3) - r3**2
        new_vertices[i, 5, 0] = r0*(r0 + 2*r2 + 2*r3) + r1*(r1 - 2*r2 + 2*r3) - r2**2 - r3**2
        new_vertices[i, 5, 1] = r0*(-r0 - 2*r1 + 2*r3) + r1*(r1 + 2*r2) + r2*(-r2 + 2*r3) + r3**2
        new_vertices[i, 5, 2] = r0*(r0 - 2*r1 - 2*r2) + r1*(-r1 + 2*r3) + r2*(-r2 - 2*r3) + r3**2
        new_vertices[i, 6, 0] = r0*(r0 - 2*r2 - 2*r3) + r1*(r1 + 2*r2 - 2*r3) - r2**2 - r3**2
        new_vertices[i, 6, 1] = r0*(r0 + 2*r1 + 2*r3) + r1*(-r1 + 2*r2) + r2*(r2 - 2*r3) - r3**2
        new_vertices[i, 6, 2] = r0*(-r0 + 2*r1 - 2*r2) + r1*(r1 + 2*r3) + r2*(r2 + 2*r3) - r3**2
        new_vertices[i, 7, 0] = r0*(r0 + 2*r2 - 2*r3) + r1*(r1 + 2*r2 + 2*r3) - r2**2 - r3**2
        new_vertices[i, 7, 1] = r0*(r0 - 2*r1 + 2*r3) + r1*(-r1 + 2*r2) + r2*(r2 + 2*r3) - r3**2
        new_vertices[i, 7, 2] = r0*(r0 + 2*r1 - 2*r2) + r1*(-r1 + 2*r3) + r2*(-r2 + 2*r3) + r3**2
    return new_vertices
```

(Note: You should test this before use, in case I made some silly mistake when generating these expressions.)  Presumably, you'll then want to scale these cubes to the appropriate sizes, and translate them as needed — which can easily be done with the usual numpy methods, or can be coded into the above function for greater efficiency.  Even with all the miracles of BLAS-accelerated matrix multiplication, this approach looks ~5 times faster.

I still don't understand how or why you're using `from_rotation_matrix`, of course.

**Comment by Nicholas-Mitchell** [comment]
Thanks for the explanations, @moble 

Your function does work and I could extend it to perform the scaling and translation of each object's 8 vertices. Comparing to my original numpy-only implementation, it is ~50x faster!
I have, however spotted an issue in that the results are not quite the same (not just talking about precision) - several objects seem to have rotations that are off by 90 degrees. I am trying to understand your function so I can debug further. By simply plotting the points from a single object, I have so far understood that you build up each object's bounding box in a certain order, as per this image, assuming the red face is the front of the object:

<img src="https://user-images.githubusercontent.com/12820901/96045908-eff7ff00-0e72-11eb-9c8e-1c67ea9b69ac.png" width="200" />

But could you please explain for example how one single vertex is constructed (e.g. point 7 in the diagram)? E.g. how do I read/understand the directions in terms of x-y-z; is it based on the sign of the four components (here `+ + - -`)?

```python
new_vertices[i, 7, 0] = r0*(r0 + 2*r2 - 2*r3) + r1*(r1 + 2*r2 + 2*r3) - r2**2 - r3**2
```

---


I am able to rotate the displacement vectors (i.e. the centroids) using the rotation matrix approach, which I also use for an entire point-cloud, containing e.g 100,000 points). The final part that I need to wrap my head around is rotating the "translated" rotation vectors of the objects around the same axis. As a concrete example, rotating the scene by positive `theta` about the z-axis, moving from `q` to `q'` (I apologise for the robotics coordinate system):

<img src="https://user-images.githubusercontent.com/12820901/96051090-102bbc00-0e7b-11eb-8648-4b42d972afe7.png" width="275" />

I understand the rotation quaternion would be of the form: `(cos(theta/2), 0, 0, sin(theta/2)`, but not sure how to apply this to the quaternion that isn't centered on our origin. (I've been using [this cool tool](https://eater.net/quaternions/video/intro) to try to _grok_ quaternions in general). I have the intuition that the angle between `q` and `q'` should be identical (i.e. `= theta`), but cannot convince myself of it. Is it covered in all cases by your example: `orientations[:] = (rotation * orientation_rotors).components`?

[For that last expression, the result of `rotation * orientation_rotors` is an `np.ndarray`, so no `components` attribute - I had to use `quaternion.as_float_array()` instead]

---

Regarding `from_rotation_matrix`: I was rotating the entire scene in one go e.g. about the z-axis (representing objects as 8-point boxes), then wanting to compute a rotation matrix for each object based on its final position/orientation, and finally convert that into a quaternion representation.

**Comment by moble** [comment]
> But could you please explain for example how one single vertex is constructed (e.g. point 7 in the diagram)? E.g. how do I read/understand the directions in terms of x-y-z; is it based on the sign of the four components (here `+ + - -`)?

The last index represents the x, y, z coordinates in that order:

```python
        new_vertices[i, 7, 0] = ...  # x-coordinate of vertex 7
        new_vertices[i, 7, 1] = ...  # y-coordinate of vertex 7
        new_vertices[i, 7, 2] = ...  # z-coordinate of vertex 7
```

The formula I used (the terms represented by `...` here) should be equivalent to this:

```python
vertex_7 = quaternion.quaternion(0, 1, 1, 1)
rotor = quaternion.quaternion(r0, r1, r2, r3)
new_vertex_7 = rotor * vertex_7 * rotor.conjugate()
```

Here, `vertex_7` is a quaternion representing the vector, so the first ("scalar") part is 0, and the following components are just the x, y, z components (in that order) of the vector.

---

Just as you can compose rotations by multiplying the rotation matrices that represent them, you can also multiply the quaternions that represent them.  So just multiply the quaternion representing the orientation of the cube by the quaternion representing the rotation of the scene.  Just be sure to put the second rotation on the left.  So if cube `i` has an orientation given by the unit quaternion `r_i` and you're rotating the entire scene by a quaternion `R`, the rotated version of cube `i` will have orientation `R * r_i`.

To see this, it might help to write down what you're doing mathematically.  Before rotation, for cube <i>i</i>, some particular vertex is moved from its original "reference" position <i>v</i> to its new position 

> <i>v<sub>i</sub> = r<sub>i</sub>&thinsp;v&thinsp;r<sub>i</sub><sup>-1</sup> + c<sub>i</sub> </i>

Now, you rotate the whole scene by <i>R</i> and the vertex goes to <i>w<sub>i</sub></i>, which is given by

> <i>w<sub>i</sub> = R&thinsp; v<sub>i</sub> &thinsp; R<sup>-1</sup></i>

Now we plug in the expression for <i>v<sub>i</sub></i> and see that this is mathematically equivalent to

> <i>w<sub>i</sub> = R&thinsp; (r<sub>i</sub>&thinsp;v&thinsp;r<sub>i</sub><sup>-1</sup> + c<sub>i</sub>) &thinsp; R<sup>-1</sup></i>

Using distributivity, we can rewrite this as

> <i>w<sub>i</sub> = R&thinsp; r<sub>i</sub>&thinsp;v&thinsp;r<sub>i</sub><sup>-1</sup>&thinsp; R<sup>-1</sup> + R&thinsp; c<sub>i</sub>&thinsp; R<sup>-1</sup> </i>

And finally, using associativity and the rule for inverses of products, this is the same as

> <i>w<sub>i</sub> = (R&thinsp; r<sub>i</sub>) &thinsp;v&thinsp; (R&thinsp; r<sub>i</sub>)<sup>-1</sup> + R&thinsp; c<sub>i</sub>&thinsp; R<sup>-1</sup> </i>

So you see that the vertex in the rotated scene is given by rotating the original centroid by <i>R</i> and representing the final orientation by <i>R&thinsp; r<sub>i</sub></i>.  There's no need to use `from_rotation_matrix`; just multiply the quaternions.

**Comment by Nicholas-Mitchell** [comment]
Thank you for the example derivation on the composition of quaternions - that is much clearer.

I have narrowed down that the rotations and unit 8-point boxes are being generated correctly (i.e. confirming no "_silly mistake_" or typo on your part :)), but as that expression represents the single vertex's _x-coordinate_ with the object already in the correct orientation, I cannot then simply multiply by the object's dimensions in order to scale it. Translation using the centroid's coordinates of course works as expected. So in order to know how to scale correctly, I need to understand the multiplications, and so know where I should put the scaling.

I tried going through the steps with the real example of `vertex_7`, performing the multiplications by hand using the [Hamilton product](https://en.wikipedia.org/wiki/Quaternion#Hamilton_product) or the scalar-vector parts method, in order to understand the expressions you gave. Unfortunately I cannot arrive at the solution you gave. Using the given values as above (and mixing some notation):

> vertex_7 = (0, 1i, 1j, 1k)
> rotor = (r0, r1, r2, r3)
> new_vertex_7 = rotor * vertex_7 * rotor.conjugate()

such that

> new_vertex_7 = (r0, r1, r2, r3) * (0, 1i, 1j, 1k) * (r0, -r1, -r2, -r3)

I cannot arrive at your solution, namely:

> new_vertices[i, 7, 0] = r0*(r0 + 2*r2 - 2*r3) + r1*(r1 + 2*r2 + 2*r3) - r2**2 - r3**2

I would assume from that, you have the penultimate step as:

> (r0, r1, r2, r3)(r0 + 2*r2 - 2*r3, r0 + 2*r2 - 2*r3, r2, r3)

or am I missing a trick?

**Comment by moble** [comment]
> I would assume from that, you have the penultimate step as:
> 
> ```python
> (r0, r1, r2, r3)(r0 + 2*r2 - 2*r3, r0 + 2*r2 - 2*r3, r2, r3)
> ```

No, that's not how I did it.  It looks like you're working backwards from my solution, and doing it in a way that won't work for the other terms; try working forwards until you get the final expression of your own, and then re-group the terms.  Specifically, I put the final expressions in ["Horner form"](https://en.wikipedia.org/wiki/Horner%27s_method), which is a more efficient way to evaluate polynomials.

Also, your notation for the original vertex is inconsistent, and I'm wondering if you're doubling some operations.  You write `(0, 1i, 1j, 1k)`, where it should be `(0, 1, 1, 1)`.

**Comment by Nicholas-Mitchell** [comment]
I had indeed tried to go forwards from the beginning, trying both Hamilton product and the scalar-vector methods:

![image](https://user-images.githubusercontent.com/12820901/96392454-691d8c00-11bc-11eb-9c3f-64edc29f33a2.png)

However I never get anything that could be reduced to be equivalent to your expression. E.g. for `new_vertices[i, 7, 0]`,
I get the four squared terms `r0**2`, `r1**2`, `r2**2` & `r3**2`, but also a `2*-r0*r1` term, which must be wrong as it doesn't appear in your answer, even once expanded. I think even discounting the Horner form, I clearly don't understand all the required quaternion arithmetic rules.

Do you have any references for this kind of calculation, or would you mind showing me how you did it? I have found plenty of examples for multiplying out two quaternions, but not a step by step example of a rotation such as `q*(0, 1, 1, 1)*q'`.

**Comment by moble** [comment]
Defining r⃗=(r1, r2, r3) and v⃗=(1, 1, 1), we can use the scalar-vector method:

```
v⃗' = (r0, r⃗)(0, v⃗)(r0, -r⃗)
   = (r0, r⃗)(v⃗·r⃗, r0 v⃗ - v⃗×r⃗)
   = r0(v⃗·r⃗, r0 v⃗ - v⃗×r⃗) + r⃗(v⃗·r⃗, r0 v⃗ - v⃗×r⃗)
   = (r0 v⃗·r⃗, r0² v⃗ - r0 v⃗×r⃗) + (-r0 v⃗·r⃗, v⃗·r⃗ r⃗ + r0 r⃗×v⃗ - r⃗×(v⃗×r⃗))
   = (0, r0² v⃗ - 2 r0 v⃗×r⃗ + v⃗·r⃗ r⃗ - r⃗×(v⃗×r⃗))
   = (0, r0² v⃗ - 2 r0 v⃗×r⃗ + v⃗·r⃗ r⃗ - r⃗² v⃗ + r⃗·v⃗ r⃗)
   = (0, r0² v⃗ - r⃗² v⃗ - 2 r0 v⃗×r⃗ + 2 v⃗·r⃗ r⃗)
```

Now, we can drop the scalar part and start substituting components:
```
v⃗' = r0² v⃗ - r⃗² v⃗ - 2 r0 v⃗×r⃗ + 2 v⃗·r⃗ r⃗
   = (r0² - r⃗²) v⃗ - 2 r0 v⃗×r⃗ + 2 (r1+r2+r3) r⃗
   = (r0² - r1² - r2² - r3²) v⃗ + 2 r0 r⃗×v⃗ + 2 (r1+r2+r3) r⃗
   = (r0² - r1² - r2² - r3²) (1, 1, 1) + 2 r0 (r2-r3, r3-r1, r1-r2) + 2 (r1+r2+r3) (r1, r2, r3)
```

So, for example, the `x` component that you're talking about is
```
(r0² - r1² - r2² - r3²) + 2 r0 (r2-r3) + 2 (r1+r2+r3) (r1)
= r0 (r0 + 2 r2 - 2 r3) + r1 (r1 + 2 r2 + 2 r3) - r2² - r3²
```
which is the expression I used.

**Comment by moble** [comment]
Oh, sorry about the rendering.  Those are all just supposed to be arrows.
